internal/controllers/topology/cluster/structuredmerge/patchhelper.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package structuredmerge
+
+import (
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/cluster-api/internal/contract"
+)
+
+const topologyManager = "topology"
+
+type serverSidePatchHelper struct {
+	client         client.Client
+	modified       *unstructured.Unstructured
+	hasChanges     bool
+	hasSpecChanges bool
+}
+
+// NewServerSidePatchHelper returns a new PatchHelper using server side apply.
+func NewServerSidePatchHelper(scheme *runtime.Scheme, original, modified client.Object, c client.Client, opts ...HelperOption) (PatchHelper, error) {
+	helperOptions := &HelperOptions{}
+	helperOptions = helperOptions.ApplyOptions(opts)
+	helperOptions.allowedPaths = []contract.Path{
+		// apiVersion, kind, name and namespace are required field for a server side apply intent.
+		{"apiVersion"},
+		{"kind"},
+		{"metadata", "name"},
+		{"metadata", "namespace"},
+		// the topology controller controls/has an opinion for labels, annotation and spec only.
+		{"metadata", "labels"},
+		{"metadata", "annotations"},
+		{"spec"},
+	}
+
+	// If required, convert the original and modified objects to unstructured and filter out all the information
+	// not relevant for the topology controller.
+
+	var originalUnstructured *unstructured.Unstructured
+	if original != nil {
+		originalUnstructured = &unstructured.Unstructured{}
+		switch original.(type) {
+		case *unstructured.Unstructured:
+			originalUnstructured = original.DeepCopyObject().(*unstructured.Unstructured)
```

Comment:
Using deepcopy here breaks for empty lists and maps:
```go
	t.Run("No-op on unstructured object having empty map[string]interface in spec", func(t *testing.T) {
		g := NewWithT(t)
		
		// Second test object having an empty map and slice as value
		obj2 := builder.InfrastructureClusterTemplate(ns.Name, "obj2").WithSpecFields(map[string]interface{}{
			"spec.emptyMap":   map[string]interface{}{},
			"spec.emptySlice": []interface{}{}, // this field is then explicitly ignored by the patch helper
		}).Build()

		// create new object having an empty map[string]interface in spec and a copy of it for further testing
		original := obj2.DeepCopy()
		modified := obj2.DeepCopy()

		// Create the object using server side apply
		g.Expect(env.PatchAndWait(ctx, original, client.FieldOwner("topology"))).To(Succeed())
		// Get created object to have managed fields
		g.Expect(env.GetAPIReader().Get(ctx, client.ObjectKeyFromObject(original), original)).To(Succeed())

		// Create a patch helper for a modified object with which has no changes.
		p0, err := NewServerSidePatchHelper(fakeScheme, original, modified, env.GetClient())
		g.Expect(err).ToNot(HaveOccurred())
		g.Expect(p0.HasChanges()).To(BeFalse())
		g.Expect(p0.HasSpecChanges()).To(BeFalse())
	})
```

DeeopCopy would remove the whole `spec` in this case and the `dryRunPatch(...)` will then detect it as changes because in managedFields there is a reference to the empty slice/map:

```

Example of the previously created object:

```sh
❯ k get GenericInfrastructureClusterTemplate -n ssa-95jv2   obj2 -o json --show-managed-fields
```
```json
{
    "apiVersion": "infrastructure.cluster.x-k8s.io/v1beta1",
    "kind": "GenericInfrastructureClusterTemplate",
    "metadata": {
        "creationTimestamp": "2022-05-12T12:57:25Z",
        "generation": 1,
        "managedFields": [
            {
                "apiVersion": "infrastructure.cluster.x-k8s.io/v1beta1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:spec": {
                        "f:emptyMap": {},
                        "f:emptySlice": {},
                        "f:template": {
                            ".": {},
                            "f:spec": {}
                        }
                    }
                },
                "manager": "topology",
                "operation": "Apply",
                "time": "2022-05-12T12:57:25Z"
            }
        ],
        "name": "obj2",
        "namespace": "ssa-95jv2",
        "resourceVersion": "295",
        "uid": "f1e14b2e-10ba-4d66-8435-24d7111ba0b8"
    },
    "spec": {
        "emptyMap": {},
        "emptySlice": [],
        "template": {
            "spec": {}
        }
    }
}
```

Reply:
I was wrong on this one. It is filterIntent which drops empty maps here.